File: cplusplus/ConditionGroup.cpp
//
// Created by yaoxu on 2021/8/23.
//

#include "ConditionGroup.h"

void ConditionGroup::BracketMatch(const std::string &source, int &i) {
    int matchTimes = 0;
    unsigned int length = source.length();
    for (; i < length; i++) {
        char ch = source[i];
        switch (ch) {
            case '(':
                matchTimes++;
                break;
            case ')':
                matchTimes--;
                if (matchTimes == 0) return;
                break;
            default:
                break;
        }
    }
}

void ConditionGroup::GenerateConditionGroup(const std::string & source, std::vector<ConditionBlock> &blockList, int depth)
{
    EnumConditionGroupToken oldToken = EnumConditionGroupToken::None;
    EnumConditionGroupToken newToken = oldToken;
    unsigned int length = source.length();
    std::string temp;
    for (int i = 0; i < length; i++)
    {
        char ch = source[i];

        // 忽略字符
        if (IsIgnoreChar(ch))
        {
            newToken = EnumConditionGroupToken::None;
        }
        else if (ch == '(')
        {
            newToken = EnumConditionGroupToken::BracketLeft;
        }
        else if (ch == '&')
        {
            newToken = EnumConditionGroupToken::And;
        }
        else if (ch == '|')
        {
            newToken = EnumConditionGroupToken::Or;
        }
        else
        {
            newToken = EnumConditionGroupToken::Content;
        }

        if (oldToken != newToken)
        {
            if (temp.length() > 0)
            {
                ConditionBlock obj;
                obj.token = oldToken;
                obj.data = temp;
                obj.depth = depth;
                blockList.push_back(obj);
                temp = "";
            }

            oldToken = newToken;
        }

        switch (newToken) {
            case EnumConditionGroupToken::BracketLeft: {
                int bracketL = i;
                int bracketR = i;
                BracketMatch(source, bracketR);

                if (bracketL < bracketR) {
                    ConditionBlock objL;
                    objL.token = EnumConditionGroupToken::BracketLeft;
                    objL.data = "(";
                    objL.depth = depth;
                    blockList.push_back(objL);

                    std::string str = source.substr(bracketL + 1, bracketR - bracketL - 1);
                    GenerateConditionGroup(str, blockList, depth + 1);

                    ConditionBlock objR;
                    objR.token = EnumConditionGroupToken::BracketRight;
                    objR.data = ")";
                    objR.depth = depth;
                    blockList.push_back(objR);

                    i = bracketR;
                    temp = "";
                }
            }
            break;
            default: {
                if (newToken != EnumConditionGroupToken::None)
                    temp += ch;
            }
            break;
        }
    }

    if (temp.length() > 0)
    {
        if (oldToken == EnumConditionGroupToken::Content)
        {
            ConditionBlock obj;
            obj.token = oldToken;
            obj.data = temp;
            obj.depth = depth;
            blockList.push_back(obj);
        }
    }
}

bool ConditionGroup::ProxyCondition(const std::string &source, bool isPrompt) {
        if (isPrompt)
        {
            printf("Start Logic Check: %s\n", source.c_str());
        }

        std::string str = source.substr(0, source.length() - 1);
        if (str == "false")
            return false;
        else
            return  true;
        // return Convert.ToBoolean(str);
        // return Convert.ToBoolean(source);
    }

    bool ConditionGroup::IsOperateCondition(const std::vector<ConditionBlock> &blockList, int depth, bool isPrompt) {
    bool finalResult = false;
    EnumConditionGroupToken currentSymbol = EnumConditionGroupToken::None;
    for (int i = 0; i < blockList.size(); i++) {
        const ConditionBlock &block = blockList[i];
        if (block.depth == depth) {
            switch (block.token) {
                case EnumConditionGroupToken::And:
                    case EnumConditionGroupToken::Or:
                        currentSymbol = block.token;
                        break;
                        case EnumConditionGroupToken::Content:
                        {
                            if (currentSymbol != EnumConditionGroupToken::None) {
                                switch (currentSymbol) {
                                    case EnumConditionGroupToken::And:
                                        finalResult = finalResult && ProxyCondition(block.data, isPrompt);
                                        break;
                                        case EnumConditionGroupToken::Or:
                                            finalResult = finalResult || ProxyCondition(block.data, isPrompt);
                                            break;
                                            default:
                                                break;
                                }
                            } else {
                                finalResult = ProxyCondition(block.data, isPrompt);
                            }
                        }
                        break;
                        case EnumConditionGroupToken::BracketLeft:
                        {
                            int bracketL = i;
                            int bracketR = bracketL;
                            for (; bracketR < blockList.size(); bracketR++) {
                                const ConditionBlock &temp = blockList[bracketR];
                                if (temp.token == EnumConditionGroupToken::BracketRight && temp.depth == depth)
                                    break;
                            }

                            std::vector<ConditionBlock> bracketList(blockList.begin() + (bracketL + 1), blockList.begin() + bracketR);
                            finalResult = finalResult || IsOperateCondition(bracketList, depth + 1, isPrompt);
                            i = bracketR;
                        }
                        break;
                        default:
                            break;
            }
        }
    }

    return finalResult;
}

bool ConditionGroup::IsMatchConditionGroup(bool isPrompt) {
    if (!conditionBlocks.empty())
    {
        return IsOperateCondition(conditionBlocks, 0, isPrompt);
    }

    return true;
}

bool ConditionGroup::Check(const std::string& source, bool isPrompt) {
    conditionBlocks.clear();
    conditionHashSet.clear();
    GenerateConditionGroup(source, conditionBlocks);

    for (auto pos = conditionBlocks.begin(); pos != conditionBlocks.end(); ++pos)
    {
        if (pos->token == EnumConditionGroupToken::Content)
            conditionHashSet.insert(pos->data);
    }

    bool result = IsMatchConditionGroup(isPrompt);
    printf("%s =====> %d\n", source.c_str(), result);
    return result;
}

bool ConditionGroup::IsIgnoreChar(char ch) {
    return ch == ' ' || ch == '\n';
}

File: python/condition_group.py
from enum import Enum

class EnumConditionGroupToken(Enum):
    '''条件组词法Token枚举'''
    Undefined = "忽略空格、换行符"
    BracketLeft = "左括号"
    BracketRight = "右括号"
    BracketContent = "括号内容"
    Or = "或运算"
    And = "且运算"
    Content = "具体条件内容"

class ConditionGroupToken():
    def __init__(self, token:EnumConditionGroupToken, source:str, depth:int) -> None:
        self.token = token
        self.source = source
        self.depth = depth

class ConditionGroup():
    def isIgnoreChar(self, ch:str) -> bool:
        if ch == ' ' or ch == '\n':
            return True
        return False

    # 测试例子，可以重载该方法
    def proxyCondition(self, source:str, isPromt:bool) -> bool:
        temp = source[0:-1]
        if temp == "true":
            return True
        else:
            return False

    def isConditionCmd(self, cmd:ConditionGroupToken, depth:int) -> bool:
        if cmd.token == EnumConditionGroupToken.Content:
            return self.proxyCondition(cmd.source, True)
        elif cmd.token == EnumConditionGroupToken.BracketContent:
            return self.directCheckConditionGroup(cmd.source,   depth + 1)

    def directCheckConditionGroup(self, source:str, depth:int = 0) -> bool:
        result:list[ConditionGroupToken] = []
        oldToken = EnumConditionGroupToken.Undefined
        newToken = oldToken

        length = len(source)
        bracketMathTimes = 0
        temp = ""
        for i in range(length):
            ch = source[i]

            # 忽略字符
            if self.isIgnoreChar(ch) and bracketMathTimes == 0:
                newToken = EnumConditionGroupToken.Undefined
            elif ch == '(':
                if bracketMathTimes == 0:
                    newToken = EnumConditionGroupToken.BracketLeft
                else:
                    newToken = EnumConditionGroupToken.BracketContent
                bracketMathTimes += 1
            elif ch == ')' and bracketMathTimes > 0:
                bracketMathTimes -= 1
                if bracketMathTimes == 0:
                    newToken = EnumConditionGroupToken.BracketRight
                else:
                    newToken = EnumConditionGroupToken.BracketContent
            elif ch == '&' and bracketMathTimes == 0:
                newToken = EnumConditionGroupToken.And
            elif ch == '|' and bracketMathTimes == 0:
                newToken = EnumConditionGroupToken.Or
            elif bracketMathTimes > 0:
                newToken = EnumConditionGroupToken.BracketContent
            else:
                newToken = EnumConditionGroupToken.Content

            if newToken != oldToken:
                if len(temp) > 0:
                    if oldToken != EnumConditionGroupToken.BracketLeft and oldToken != EnumConditionGroupToken.BracketRight:
                        result.append(ConditionGroupToken(oldToken, temp, depth))
                    temp = ""
                oldToken = newToken

            if newToken != EnumConditionGroupToken.Undefined:
                temp += ch

        if len(temp) > 0:
            if oldToken == EnumConditionGroupToken.Content:
                result.append(ConditionGroupToken(oldToken, temp, depth))

        # 检测条件是否满足
        finalResult = self.isConditionCmd(result[0], depth)

        for i in range(len(result) - 1):
            cmd = result[i]
            if cmd.token == EnumConditionGroupToken.And:
                finalResult = finalResult and self.isConditionCmd(result[i + 1], depth)
            elif cmd.token == EnumConditionGroupToken.Or:
                finalResult = finalResult or self.isConditionCmd(result[i + 1], depth)

        return finalResult

    def directCheck(self, source:str) -> bool:
        return self.directCheckConditionGroup(source)